src/odemis/acq/acqmng.py

```diff
@@ -83,6 +84,189 @@ def acquire(streams, settings_obs=None):
+    """
+    This class represents an acquisition task for a zstack of data.
+    """
+
+    def __init__(self, future, streams, zlevels, settings_obs):
+        """
+        The class constructor
+        future (ProgressiveFuture): the future that will execute the task asynchronously
+        streams (list of Stream): the streams to be acquired
+        zlevels (dict Stream -> list of floats): a dictionary containing the streams and the
+        corresponding lists of actuator z positions
+        settings_obs (SettingsObserver):
+        """
+        self._main_future = future
+        self._future_state = None
+        self._streams = sortStreams(streams)
+        self._zpos = next((l for l in zlevels.values() if len(l) > 1), None)
+        self._zlevels = zlevels
+        self._settings_obs = settings_obs
+        self._zstep_duration = self._estimate_zstep_duration()
+
+    def cancel(self, _):
+        """
+        Called when the user presses the button "cancel" and the future is still running.
+        (but there might  be some delay (not instantly called) )
+        return (Boolean): True if canceled, False otherwise
+        """
+        self._future_state = CANCELLED
+        # cancel the sub-futures
+        self._actuator_f.cancel()
+        self._single_acqui_f.cancel()
+        return True
+
+    def _estimate_zstep_duration(self):
+        """
+        return (float > 0): estimated time (in s) that it takes to move the focus actuator
+          by one step.
+        """
+        # get the focuser of any of the FM streams
+        focuser = next(
+            (s.focuser for s in self._streams if isinstance(s, FluoStream)), None
+        )
+        speed = None
+        if focuser is not None:
+            if model.hasVA(focuser, "speed"):
+                speed = focuser.speed.value.get("z", None)
+        if speed is None:
+            speed = 10e-6  # m/s, pessimistic
+        zstep = 0
+        if self._zpos:
+            zstep = abs(self._zpos[0] - self._zpos[1])
+        return estimateMoveDuration(zstep, speed, 0.01)
+
+    def estimate_total_duration(self):
+        """
+        Estimates the total time for the zstack acquisition.
+        return (Float > 0): the estimated time
+        """
+        acqui_time = 0.0
+        zstep_time = 0
+        nr_FM_streams = 0
+        for s in self._streams:
+            acqui_time += s.estimateAcquisitionTime() * len(self._zlevels[s])
+            if isinstance(s, FluoStream):
+                nr_FM_streams += 1
+        if self._zpos:
+            zstep_time = self._zstep_duration * (len(self._zpos) - 1) * nr_FM_streams
+        return acqui_time + zstep_time
+
+    def _construct_zcube(self, data_arrays):
+        """
+        Create 3D data array from the 2D acquired data arrays 
+        data_arrays (list of DataArray): a list of acquired images at different z levels.
+        return (DataArray): a data array embedding all the data acquired at all z levels.
+        """
```

Comment:
This code is nearly identical as `_tiledacq._assembleZCube()`. Please move the function to `util.img.assembleZCube()`. In general, I think the code from tiledacq is safer, so I'd suggest to use it as basis.

Reply:
Okay, i will make separate PR for this change, as the changes in the _tiledacq.py is not relevant to enzel acquisition panel